platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/widgets/CompassView.java

```diff
@@ -57,105 +57,98 @@ private void initialize(Context context) {
-                    @Override
-                    public void run() {
-                        post(new Runnable() {
-                            @Override
-                            public void run() {
-                                setAlpha(1.0f);
-                                fadeAnimator = ViewCompat.animate(CompassView.this).alpha(0.0f).setDuration(1000).withLayer();
-                                fadeAnimator.setListener(new ViewPropertyAnimatorListenerAdapter() {
-                                    @Override
-                                    public void onAnimationEnd(View view) {
-                                        setVisibility(View.INVISIBLE);
-                                        northTimer = null;
-                                    }
-                                });
-                            }
-                        });
-                    }
-                }, 1000);
-            }
+            postDelayed(this, 500);
+            return;
         } else {
-            if (northTimer != null) {
-                northTimer.cancel();
-                northTimer = null;
-            }
-            if (fadeAnimator != null) {
-                fadeAnimator.cancel();
-            }
+            resetAnimation();
             setAlpha(1.0f);
             setVisibility(View.VISIBLE);
         }
+
+        setRotation((float) direction);
+    }
+
+    public void setCompassFadeFacingNorth(boolean compassFadeFacingNorth) {
+        fadeCompassViewFacingNorth = compassFadeFacingNorth;
+    }
+
+    @Override
+    public void run() {
+        if (isFacingNorth() && fadeCompassViewFacingNorth) {
+            resetAnimation();
+            fadeAnimator = ViewCompat.animate(CompassView.this).alpha(0.0f).setDuration(500).withLayer();
+            fadeAnimator.setListener(new ViewPropertyAnimatorListenerAdapter() {
+                @Override
+                public void onAnimationEnd(View view) {
+                    setVisibility(View.INVISIBLE);
+                    resetAnimation();
+                }
+            });
+        }
     }
 
     static class CompassClickListener implements View.OnClickListener {
 
         private WeakReference<MapboxMap> mapboxMap;
+        private WeakReference<CompassView> compassView;
 
-        public CompassClickListener(final MapboxMap mapboxMap) {
+        CompassClickListener(final MapboxMap mapboxMap, CompassView compassView) {
             this.mapboxMap = new WeakReference<>(mapboxMap);
+            this.compassView = new WeakReference<>(compassView);
         }
 
         @Override
         public void onClick(View view) {
             final MapboxMap mapboxMap = this.mapboxMap.get();
-            if (mapboxMap != null) {
+            final CompassView compassView = this.compassView.get();
+            if (mapboxMap != null && compassView != null) {
                 mapboxMap.resetNorth();
+                compassView.postDelayed(compassView, 650);
             }
```

Comment:
The other one is slightly longer since the MapView needs to reset to north before the fading occurs. I'm going to replace the magic constants with actual constants.